src/cluster/websocket/WebSocket.h

```diff
@@ -0,0 +1,301 @@
+
+    /**
+     * @return The WebSocket client's remote port.
+     */
+    virtual int getRemotePort() = 0;
+
+    /**
+     * Store the client's subscriptions as "not active".
+     */
+    void SetSubscriptions(const std::vector<std::string>& topic_prefixes);
+
+    /**
+     * @return The client's subscriptions.
+     */
+    const std::vector<std::string>& GetSubscriptions() const;
+
+    /**
+     * Store the client's subscriptions as "not active".
+     */
+    void SetSubscriptionActive(std::string& topic_prefix);
+
+    /**
+     * @return true if all subscriptions have an active status.
+     */
+    bool AllSubscriptionsActive() const;
+
+private:
+    bool acked = false;
+    std::map<std::string, bool> subscriptions_state;
+    std::vector<std::string> subscriptions;
+};
+
+// An new WebSocket client connected. Client is locally identified by `id`.
+struct WebSocketOpen {
+    std::string id;
+    std::shared_ptr<WebSocketClient> wsc;
+};
+
+// A WebSocket client disconnected.
+struct WebSocketClose {
+    std::string id;
+};
+
+// A WebSocket client send a message.
+struct WebSocketMessage {
+    std::string id;
+    std::string msg;
+};
+
+// Produced internally when a WebSocket client's
+// subscription has completed.
+struct WebSocketSubscribeFinished {
+    std::string id;
+    std::string topic_prefix;
+};
+
+using WebSocketEvent = std::variant<WebSocketOpen, WebSocketSubscribeFinished, WebSocketClose, WebSocketMessage>;
+
+struct WebSocketSendReply {
+    std::shared_ptr<WebSocketClient> wsc;
+    std::string msg;
+};
+
+struct WebSocketCloseReply {
+    std::shared_ptr<WebSocketClient> wsc;
+    uint16_t code = 1000;
+    std::string reason = "Normal closure";
+};
+
+using WebSocketReply = std::variant<WebSocketSendReply, WebSocketCloseReply>;
+
+
+class ReplyMsgThread;
+
+/**
+ * Handle events produced by WebSocket clients.
+ *
+ * Any thread may call QueueForProcessing(). Process() runs
+ * on Zeek's main thread.
+ */
+class WebSocketEventDispatcher {
+public:
+    WebSocketEventDispatcher();
+
+    ~WebSocketEventDispatcher();
+
+    /**
+     * Called shutting down a WebSocket server.
+     */
+    void Terminate();
+
+    /**
+     * Queue the given WebSocket event to be processed on Zeek's main loop.
+     *
+     * @param work The WebSocket event to process.
+     */
+    void QueueForProcessing(WebSocketEvent&& event);
+
+    /**
+     * Send a reply to the given websocket client.
+     *
+     * The dispatcher has an internal thread for serializing
+     * and sending out the event.
+     */
+    void QueueReply(WebSocketReply&& reply);
+
+private:
+    /**
+     * Main processing function of the dispatcher.
+     *
+     * This runs on Zeek's main thread.
+     */
+    void Process(WebSocketEvent&& event);
+
+    void Process(WebSocketOpen& open);
+    void Process(WebSocketSubscribeFinished& fin);
+    void Process(WebSocketMessage& msg);
+    void Process(WebSocketClose& close);
```

Comment:
Not sure why we need mutable refs, maybe pass like we pass `WebsocketEvent` above?

```suggestion
    void Process(WebSocketOpen&& open);
    void Process(WebSocketSubscribeFinished&& fin);
    void Process(WebSocketMessage&& msg);
    void Process(WebSocketClose&& close);
```

Reply:
That change results in a stack overflow. Happy to debug with you :-)